src/main/java/ca/corefacility/bioinformatics/irida/ria/web/services/UIMetadataImportService.java

```diff
@@ -1,172 +1,241 @@
+	private final MetadataTemplateService metadataTemplateService;
+	private final ProjectControllerUtils projectControllerUtils;
+	private final UIMetadataFileImportService metadataFileImportService;
+
 	@Autowired
-	public UIMetadataImportService() {
+	public UIMetadataImportService(MessageSource messageSource, ProjectService projectService,
+			SampleService sampleService, MetadataTemplateService metadataTemplateService,
+			ProjectControllerUtils projectControllerUtils, UIMetadataFileImportService metadataFileImportService) {
+		this.messageSource = messageSource;
+		this.projectService = projectService;
+		this.sampleService = sampleService;
+		this.metadataTemplateService = metadataTemplateService;
+		this.projectControllerUtils = projectControllerUtils;
+		this.metadataFileImportService = metadataFileImportService;
 	}
 
 	/**
-	 * Parse metadata from an csv file.
+	 * Upload CSV or Excel file containing sample metadata and extract the headers.  The file is stored in the session until
+	 * the column that corresponds to a {@link Sample} identifier has been sent.
 	 *
-	 * @param inputStream The inputStream of the csv file.
-	 * @return {@link SampleMetadataStorage} contains the metadata from file.
-	 * @throws IOException thrown if the extension does not exist.
+	 * @param session   {@link HttpSession}
+	 * @param projectId {@link Long} identifier for the current {@link Project}
+	 * @param file      {@link MultipartFile} The csv or excel file containing the metadata.
+	 * @return {@link Map} of headers and rows from the csv or excel file for the user to select the header corresponding the
+	 * {@link Sample} identifier.
 	 */
-	public SampleMetadataStorage parseCSV(InputStream inputStream) throws IOException {
+	public SampleMetadataStorage createProjectSampleMetadata(HttpSession session, long projectId, MultipartFile file) {
+		// We want to return a list of the table headers back to the UI.
 		SampleMetadataStorage storage = new SampleMetadataStorage();
-
-		CSVParser parser = CSVParser.parse(inputStream, StandardCharsets.UTF_8,
-				CSVFormat.RFC4180.withFirstRecordAsHeader()
-						.withTrim()
-						.withIgnoreEmptyLines());
-		List<Map<String, String>> rows = new ArrayList<>();
-
-		// save headers
-		Map<String, Integer> headers_set = parser.getHeaderMap();
-		List<String> headers_list = new ArrayList<>(headers_set.keySet());
-		storage.saveHeaders(headers_list);
-
-		// save data
-		for (CSVRecord row : parser) {
-			Map<String, String> rowMap = new HashMap<>();
-			for (String key : row.toMap()
-					.keySet()) {
-				String value = row.toMap()
-						.get(key);
-				rowMap.put(key, value);
+		try (InputStream inputStream = file.getInputStream()) {
+			String filename = file.getOriginalFilename();
+			String extension = Files.getFileExtension(filename);
+
+			// Check the file type
+			switch (extension) {
+			case "csv":
+				storage = metadataFileImportService.parseCSV(projectId, inputStream);
+				break;
+			case "xlsx":
```

Comment:
Does something need to be added for this case?

Reply:
No, I don't think so. For both xlsx and xls cases I want to call the parseExcel method. There's no break within the xlsx case.

Reply:
Oh I see what it's doing. Disregard :)